passcode/Source/ViewControllers/Passcode/PasscodeViewController.swift

```
@@ -0,0 +1,152 @@
+        messageLabel.numberOfLines = 0
+        messageLabel.text = "Tap \"done\" on the keyboard when you done entering."
+        stackView.addArrangedSubview(messageLabel)
+
+        #if DEBUG
+        if !self.viewModel.isSetupMode {
+            let hintLabel = UILabel()
+            hintLabel.translatesAutoresizingMaskIntoConstraints = false
+            hintLabel.textColor = UIColor.tertiaryLabel
+            hintLabel.numberOfLines = 0
+            hintLabel.text = """
+            Current passcode: \((try? PasscodeUtility.currentPasscode())?.value ?? "nil")
+            For DEBUG only. Since passcode is stored in keychain, even deleting the app won't remove it.
+            """
+            stackView.addArrangedSubview(hintLabel)
+        }
+        #endif
+
+        if self.viewModel.isSetupMode {
+            self.navigationItem.leftBarButtonItem = UIBarButtonItem(title: "Cancel", style: .plain, target: self, action: #selector(cancelButtonTouchUpInside))
+        }
+    }
+
+    @objc private func cancelButtonTouchUpInside() {
+
+        self.dismiss(animated: true) {
+            if self.viewModel.isSetupMode {
+                self.delegate?.didDismissForCreation(creationSucceed: false)
+            }
+        }
+    }
+}
+
+extension PasscodeViewController: UITextFieldDelegate {
+
+    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
+
+        guard textField == self.textField else {
+            return true
+        }
+
+        self.textField.resignFirstResponder()
+
+        if let passcode = textField.text {
+
+            switch self.viewModel.processPasscode(passcode) {
+            case .success:
+                self.errorMessageLabel.text = nil
+                switch self.viewModel.mode {
+
+                case .validation:
+                    self.dismiss(animated: true, completion: nil)
+                case .setup(let isConfirmed):
+                    if isConfirmed {
```

Hide error message first, then dismiss